pkg/db/db.go

```diff
@@ -0,0 +1,118 @@
+package db
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/jackc/pgx/v5/pgxpool"
+	"github.com/red-hat-storage/managed-fusion-fleet-reconciler/pkg/types"
+	"go.uber.org/zap"
+)
+
+func GetConnectionString(host, user, password, name string, port int) string {
+	return fmt.Sprintf(
+		"host=%s port=%d user=%s password=%s dbname=%s sslmode=disable",
+		host,
+		port,
+		user,
+		password,
+		name,
+	)
+}
+
+type Database struct {
+	tables struct {
+		providers string
+	}
+	pool *pgxpool.Pool
+	// conn is used to listen for notifications and should be closed
+	conn *pgxpool.Conn
+}
+
+func NewClient(ctx context.Context, connString string, tables map[string]string) (*Database, error) {
+	pool, err := pgxpool.New(ctx, connString)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new instace of pool: %v", err)
+	}
+
+	if err := pool.Ping(ctx); err != nil {
+		return nil, fmt.Errorf("failed to ping database: %v", err)
+	}
+
+	if _, ok := tables["providers"]; !ok {
+		return nil, fmt.Errorf("incomplete table name mapping, missing a table name for \"providers\" table")
+	}
+
+	db := Database{}
+	db.pool = pool
+	db.tables.providers = tables["providers"]
+	return &db, nil
+}
+
+func (pg *Database) Close(ctx context.Context) {
+	if pg.conn != nil {
+		pg.conn.Conn().Close(ctx)
+	}
+	pg.pool.Close()
+}
+
+func (pg *Database) OnProvider(ctx context.Context, logger *zap.Logger, notifyExisting bool, fn func(providerName string)) error {
+	var err error
+	if pg.conn, err = pg.pool.Acquire(ctx); err != nil {
+		return fmt.Errorf("failed to acquire connection: %v", err)
+	}
+
+	query := fmt.Sprintf("LISTEN %s;", pg.tables.providers)
+	// Listen for notifications
+	if _, err = pg.conn.Exec(ctx, query); err != nil {
+		return err
+	}
+	go func() {
+		for {
+			notification, err := pg.conn.Conn().WaitForNotification(ctx)
```

Comment:
- just curious, were you able to test with providers count more than workers with no worker is available for work?
- I know in the prototype we didn't use db notifications, simply wanted to know if these gets buffered or not, a link to the doc also suffices.

Reply:
This is not tested with DB i don't see a problem with that, if you see a problem I can take a look at it now or if we get a problem in this area it can be addressed later.